src/vendor/libbacktrace_2446c66076480ce07a6bd868badcbceb3eeecc2e_debuginfod_patch.diff

```diff
@@ -0,0 +1,331 @@
+diff --git a/elf.c b/elf.c
+index fb54165..78efe22 100644
+--- a/elf.c
++++ b/elf.c
+@@ -30,6 +30,8 @@ STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+ IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ POSSIBILITY OF SUCH DAMAGE.  */
+ 
++#define _GNU_SOURCE
++
+ #include "config.h"
+ 
+ #include <errno.h>
+@@ -38,6 +40,7 @@ POSSIBILITY OF SUCH DAMAGE.  */
+ #include <sys/types.h>
+ #include <sys/stat.h>
+ #include <unistd.h>
++#include <dlfcn.h>
+ 
+ #ifdef HAVE_DL_ITERATE_PHDR
+ #include <link.h>
+@@ -45,6 +48,7 @@ POSSIBILITY OF SUCH DAMAGE.  */
+ 
+ #include "backtrace.h"
+ #include "internal.h"
++#include "debuginfod_support.h"
+ 
+ #ifndef S_ISLNK
+  #ifndef S_IFLNK
+@@ -851,12 +855,86 @@ elf_readlink (struct backtrace_state *state, const char *filename,
+ 
+ #define SYSTEM_BUILD_ID_DIR "/usr/lib/debug/.build-id/"
+ 
++typedef debuginfod_client*(*debuginfod_begin_t)(void);
++
++typedef int(*debuginfod_find_debuginfo_t)(debuginfod_client *client,
++                                            const unsigned char *build_id,
++                                            int build_id_len,
++                                            char **path);
++typedef void(*debuginfod_end_t)(debuginfod_client *client);
++
++ 
++static debuginfod_find_debuginfo_t the_debuginfod_find_debuginfo = NULL;
++static debuginfod_begin_t the_debuginfod_begin = NULL;
++static debuginfod_end_t the_debuginfod_end = NULL;
++static int debuginfod_guard = 0;
++static int debuginfod_dlopen_failed = 0;
++ 
++static int
++elf_open_debugfile_by_debuginfod (const char *buildid_data,
++			       size_t buildid_size,
++			       backtrace_error_callback error_callback,
++			       void *data)
++{
++  int ret;
++  int does_not_exist;
++
++  if (debuginfod_dlopen_failed) {
++      return -1;
++  }
++
++  if (the_debuginfod_find_debuginfo == NULL) {
++    void *handle = dlopen(DEBUGINFOD_SONAME, RTLD_LAZY);
++    if (!handle) {
++      debuginfod_dlopen_failed = 1;
++      return -1;
++    }
++
++    the_debuginfod_find_debuginfo = (debuginfod_find_debuginfo_t)dlsym(handle, "debuginfod_find_debuginfo");
++    if (!the_debuginfod_find_debuginfo) {
++      return -1;
++    }
++
++    the_debuginfod_begin = (debuginfod_begin_t)dlsym(handle, "debuginfod_begin");
++    if (!the_debuginfod_begin) {
++      return -1;
++    }
++
++    the_debuginfod_end = (debuginfod_end_t)dlsym(handle, "debuginfod_end");
++    if (!the_debuginfod_end) {
++      return -1;
++    }
++  }
++
++  char* path = NULL;
++  const unsigned char* bi = (const unsigned char*)buildid_data;
++
++  debuginfod_client* client = the_debuginfod_begin();
++  if (!client) {
++    return -1;
++  }
++
++  int fd = the_debuginfod_find_debuginfo(client, bi, buildid_size, &path);
++
++  the_debuginfod_end(client);
++
++  if (!path || fd <= 0) {
++    return -1;
++  } 
++
++  ret = backtrace_open (path, error_callback, data, &does_not_exist);
++
++  free(path);
```

Comment:
The docs say:
> If a find family function is successful, the resulting file is saved
> to the client cache and a file descriptor to that file is returned.
> The caller needs to `close()` this descriptor.  Otherwise, a
> negative error code is returned.

I think we need:
```suggestion
+  free(path);
+  close(fd);
```